dbt/include/sqlserver/macros/adapters/columns.sql

```diff
@@ -1,48 +1,27 @@
 {% macro sqlserver__get_columns_in_relation(relation) -%}
   {% call statement('get_columns_in_relation', fetch_result=True) %}
 
-    with
-        regular_db_cols as (
-            select
-                ordinal_position,
-                column_name,
-                data_type,
-                character_maximum_length,
-                numeric_precision,
-                numeric_scale
-            from [{{ relation.database }}].INFORMATION_SCHEMA.COLUMNS
-            where table_name = '{{ relation.identifier }}'
-              and table_schema = '{{ relation.schema }}'
-        ),
-
-        temp_db_cols as (
-            select
-                ordinal_position,
-                column_name collate database_default as column_name,
-                data_type collate database_default as data_type,
-                character_maximum_length,
-                numeric_precision,
-                numeric_scale
-            from tempdb.INFORMATION_SCHEMA.COLUMNS
-            where table_name like '{{ relation.identifier }}%'
-        ),
-
-        all_cols as (
-            select *
-            from regular_db_cols
-            union
-            select *
-            from temp_db_cols
-        )
+    with mapping as (
+        select
+            row_number() over (partition by object_name(c.object_id) order by c.column_id) as ordinal_position,
+            c.name collate database_default as column_name,
+            t.name as data_type,
+            c.max_length as character_maximum_length,
+            c.precision as numeric_precision,
+            c.scale as numeric_scale
+        from [{{ 'tempdb' if '#' in relation.identifier else relation.database }}].sys.columns c
+        inner join sys.types t
+        on c.user_type_id = t.user_type_id
+        where c.object_id = object_id('{{ 'tempdb..' ~ relation.include(database=false, schema=false) if '#' in relation.identifier else relation }}')
+    )
```

Comment:
can you help me understand what this does differently than before

Reply:
When you create a temp table, SQL Server will actually name it `#yourname_____...____somerandomvalue`
When you reference the temp table by its regular name (`#yourname`), you can find it and use it as normal. In this case, the previous version looked for the temp table using information_schema (ansi/cross-db version of sys.*) but it looked for all matches with `LIKE "#yourname%`. In a single-user/single-session scenario, you'd only get a single result. But if multiple users are working on the same database (with shared tempdb) and use temp tables with the same name (as temp tables don't have a schema), you'd get multiple matches. This is why sometimes our test failed when multiple instances were running at the same time, the result was not reliable.
This only reliable way to get all columns in *just your* temp table, without including the ones from other users is by referencing it in a normal way and let SQL Server do its magic to find back the one with the name suffix. In our case this means that you have to ref it using an object_id which is always unique. But you can't use object_id with information_schema, so I had to revert to sys.columns and sys.types to get the same information.

Reply:
1. Seems a bit faster when running multiple tests concurrently.
2. My assumption is that the override is no longer needed.